File: photo-studio/tests/test_gemini_client.py
from __future__ import annotations

from dataclasses import replace

from PIL import Image

from app.gemini_client import GeminiClient, build_instruction
from app.models import ProcessingMode
from app.similarity import structural_similarity


def test_not_configured_without_api_key(test_settings):
    client = GeminiClient(settings=test_settings)
    assert client.is_configured is False

    img = Image.new("RGB", (100, 100), (255, 255, 255))
    result = client.edit_image(img, ProcessingMode.CLEAN_PRODUCT_PHOTO, "neutral_studio")
    assert result.image is None
    assert result.is_safe is False
    assert result.error is not None


def test_configured_when_api_key_present(test_settings):
    configured_settings = replace(test_settings, gemini_api_key="fake-test-key-not-real")
    client = GeminiClient(settings=configured_settings)
    assert client.is_configured is True


def test_build_instruction_preserves_defect_language():
    instruction = build_instruction(ProcessingMode.CLEAN_PRODUCT_PHOTO, "white")
    assert "stains" in instruction.lower()
    assert "holes" in instruction.lower()
    assert "preserve" in instruction.lower()
    assert "logos" in instruction.lower()


def test_build_instruction_respects_preserve_original_background():
    instruction = build_instruction(ProcessingMode.CLEAN_PRODUCT_PHOTO, "preserve_original")
    assert "keep the original background" in instruction.lower()


def test_ssim_identical_images_is_near_one():
    img = Image.new("RGB", (200, 200), (120, 80, 40))
    score = structural_similarity(img, img)
    assert score > 0.99


def test_ssim_very_different_images_is_low():
    img_a = Image.new("RGB", (200, 200), (255, 255, 255))
    img_b = Image.new("RGB", (200, 200), (0, 0, 0))
    score = structural_similarity(img_a, img_b)
    assert score < 0.5

File: photo-studio-php/tests/test_gemini_client.php
<?php
declare(strict_types=1);

test('not configured without api key returns unsafe error result', function () {
    // This dev/test environment has no GEMINI_API_KEY in .env by default.
    if (jps_gemini_configured()) {
        return; // A real key is configured in this environment; skip.
    }
    $img = imagecreatetruecolor(100, 100);
    $result = jps_gemini_edit_image($img, 'clean_product_photo', 'neutral_studio', sys_get_temp_dir());
    assert_true($result['image'] === null);
    assert_false($result['is_safe']);
    assert_true($result['error'] !== null);
});

test('build_instruction preserves defect language', function () {
    $instruction = jps_build_gemini_instruction('clean_product_photo', 'white');
    $lower = strtolower($instruction);
    assert_true(str_contains($lower, 'stains'));
    assert_true(str_contains($lower, 'holes'));
    assert_true(str_contains($lower, 'preserve'));
    assert_true(str_contains($lower, 'logos'));
});

test('build_instruction respects preserve_original background', function () {
    $instruction = jps_build_gemini_instruction('clean_product_photo', 'preserve_original');
    assert_true(str_contains(strtolower($instruction), 'keep the original background'));
});

test('similarity of identical images is near one', function () {
    $img = imagecreatetruecolor(200, 200);
    $color = imagecolorallocate($img, 120, 80, 40);
    imagefilledrectangle($img, 0, 0, 200, 200, $color);
    $score = jps_image_similarity($img, $img);
    assert_true($score > 0.98, "expected near-1.0 similarity, got $score");
});

test('similarity of very different images is low', function () {
    $a = imagecreatetruecolor(200, 200);
    $white = imagecolorallocate($a, 255, 255, 255);
    imagefilledrectangle($a, 0, 0, 200, 200, $white);

    $b = imagecreatetruecolor(200, 200);
    $black = imagecolorallocate($b, 0, 0, 0);
    imagefilledrectangle($b, 0, 0, 200, 200, $black);

    $score = jps_image_similarity($a, $b);
    assert_true($score < 0.3, "expected low similarity, got $score");
});
